crypto/evp/p_lib.c

```diff
@@ -317,78 +318,234 @@ int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
+        ENGINE *tmpe = NULL;
+
+        if (strtype != NULL)
+            ameth = EVP_PKEY_asn1_find_str(&tmpe, strtype, -1);
+        else if (nidtype != EVP_PKEY_NONE)
+            ameth = EVP_PKEY_asn1_find(&tmpe, nidtype);
+
+        /* If tmpe is NULL then no engine is claiming to support this type */
+        if (tmpe == NULL)
+            ameth = NULL;
+
+        ENGINE_finish(tmpe);
+    }
+
+    if (e == NULL && ameth == NULL) {
+        /*
+         * No engine is claiming to support this type, so lets see if we have
+         * a provider.
+         */
+        ctx = EVP_PKEY_CTX_new_from_name(libctx,
+                                         strtype != NULL ? strtype
+                                                         : OBJ_nid2sn(nidtype),
+                                         propq);
+        if (ctx == NULL) {
+            EVPerr(0, ERR_R_MALLOC_FAILURE);
+            goto err;
+        }
+        if (EVP_PKEY_key_fromdata_init(ctx) == 1) {
+            tmpl = OSSL_PARAM_BLD_new();
+            if (tmpl == NULL) {
+                EVPerr(0, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
+
+            if (!priv && !OSSL_PARAM_BLD_push_octet_string(tmpl,
+                                                           OSSL_PKEY_PARAM_PUB_KEY,
+                                                           key, len)) {
+                EVPerr(0, EVP_R_KEY_SETUP_FAILED);
+                goto err;
+            }
+
+            if (priv && !OSSL_PARAM_BLD_push_octet_string(tmpl,
+                                                          OSSL_PKEY_PARAM_PRIV_KEY,
+                                                          key, len)) {
+                EVPerr(0, EVP_R_KEY_SETUP_FAILED);
+                goto err;
+            }
+
+            params = OSSL_PARAM_BLD_to_param(tmpl);
+            if (params == NULL) {
+                EVPerr(0, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
+
+            if (EVP_PKEY_fromdata(ctx, &pkey, params) != 1) {
+                EVPerr(0, EVP_R_KEY_SETUP_FAILED);
+                goto err;
+            }
+
+            OSSL_PARAM_BLD_free(tmpl);
+            OSSL_PARAM_BLD_free_params(params);
+            EVP_PKEY_CTX_free(ctx);
+
+            return pkey;
+        }
+        /* else not supported so fallback to legacy */
+    }
+
+    /* Legacy code path */
+
+    pkey = EVP_PKEY_new();
+    if (pkey == NULL) {
+        EVPerr(0, ERR_R_MALLOC_FAILURE);
         goto err;
     }
 
-    if (ret->ameth->set_priv_key == NULL) {
-        EVPerr(EVP_F_EVP_PKEY_NEW_RAW_PRIVATE_KEY,
-               EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
+    if (!pkey_set_type(pkey, e, nidtype, strtype, -1, NULL)) {
+        /* EVPerr already called */
         goto err;
     }
 
-    if (!ret->ameth->set_priv_key(ret, priv, len)) {
-        EVPerr(EVP_F_EVP_PKEY_NEW_RAW_PRIVATE_KEY, EVP_R_KEY_SETUP_FAILED);
+    if (pkey->ameth == NULL) {
+        /* Shouldn't happen */
```

Comment:
This could be an assertion, then, i.e. an extra check that `pkey_set_type()` did its job.

Reply:
Done